r2i/edgetpu/engine.cc

```diff
@@ -0,0 +1,98 @@
+ */
+
+#include "r2i/edgetpu/engine.h"
+
+#include <edgetpu.h>
+
+namespace r2i {
+namespace edgetpu {
+
+Engine::Engine () {
+  this->state = r2i::tflite::Engine::State::STOPPED;
+  this->model = nullptr;
+  this->number_of_threads = 1;
+  this->allow_fp16 = 0;
+}
+
+Engine::~Engine () {
+  r2i::tflite::Engine::Stop();
+}
+
+void Engine::setupResolver(::tflite::ops::builtin::BuiltinOpResolver
+                           &resolver) {
+  resolver.AddCustom(::edgetpu::kCustomOp, ::edgetpu::RegisterCustomOp());
+}
+
+void Engine::setInterpreterContext() {
+  std::shared_ptr<::edgetpu::EdgeTpuContext> edgetpu_context =
+    ::edgetpu::EdgeTpuManager::GetSingleton()->OpenDevice();
+  this->interpreter->SetExternalContext(kTfLiteEdgeTpuContext,
+                                        edgetpu_context.get());
+}
+
+float *Engine::runInference(std::shared_ptr<r2i::IFrame> frame,
+                            const int &input, const int &width, const int &height, const int &channels,
+                            r2i::RuntimeError &error) {
+  auto input_tensor = this->interpreter->typed_tensor<uint8_t>(input);
+  auto input_data = (float *)frame->GetData();
+
+  if (!input_data) {
+    error.Set (RuntimeError::Code::FRAMEWORK_ERROR, "Failed to get image data");
+    return nullptr;
+  }
+
+  // Convert to fixed point
+  std::unique_ptr<uint8_t> input_data_fixed(new uint8_t(height * width *
+      channels));
+  int index;
+  for (index = 0; index < height * width * channels; index++) {
+    input_data_fixed.get()[index] = static_cast<uint8_t>(input_data[index]);
+  }
+
+  memcpy(input_tensor, input_data_fixed.get(),
+         height * width * channels * sizeof(uint8_t));
+
+  if (this->interpreter->Invoke() != kTfLiteOk) {
+    error.Set (RuntimeError::Code::FRAMEWORK_ERROR,
+               "Failed to invoke tflite!");
+    return nullptr;
+  }
+
+  const auto &output_indices = interpreter->outputs();
+  const auto *out_tensor = interpreter->tensor(output_indices[0]);
+  std::unique_ptr<float> output_data;
+
+  if (out_tensor->type == kTfLiteUInt8) {
+    uint8_t *output_data_fixed = interpreter->typed_output_tensor<uint8_t>(0);
+    TfLiteIntArray *output_dims = this->interpreter->tensor(
+                                    output_indices[0])->dims;
+
+    // Convert to fixed point
+    auto output_size = GetRequiredBufferSize(output_dims);
+    output_data = std::unique_ptr<float>(new float(output_size));
+    for (index = 0; index < output_size; index++) {
+      output_data.get()[index] = static_cast<float>(output_data_fixed[index]);
+    }
+  } else if (out_tensor->type == kTfLiteFloat32) {
+    output_data = std::unique_ptr<float>(interpreter->typed_output_tensor<float>
+                                         (0));
+  } else {
+    error.Set (RuntimeError::Code::FRAMEWORK_ERROR,
+               "Output tensor has unsupported output type");
+    return nullptr;
+  }
+
+  return output_data.get();
+}
+
+} //namepsace edgetpu
```

Comment:
What are these commented lines for?

Reply:
This is just for visualize better the close of the different namespaces.